File: 084. Largest Rectangle in Histogram/Python/Largest Rectangle in Histogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        stack = []
        i = 0
        result = 0
        while i < len(heights):
            while stack and heights[i] < heights[stack[-1]]:
                index = stack.pop()
                current = heights[index] * (i - (stack[-1] if stack else -1) - 1)
                result = max(result, current)
            stack.append(i)
            i += 1
        while stack:
            index = stack.pop()
            current = heights[index] * (i - (stack[-1] if stack else -1) - 1)
            result = max(result, current)
        return result

File: 084. Largest Rectangle in Histogram/C++/Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int> &heights) {
        stack<int> sta;
        int i = 0;
        int result = 0;
        while (i < heights.size()) {
            while (!sta.empty() && heights[i] < heights[sta.top()]) {
                int index = sta.top();
                sta.pop();
                int current = heights[index] * (i - (sta.empty() ? -1 : sta.top()) - 1);
                result = max(result, current);
            }
            sta.push(i);
            i++;
        }
        while (!sta.empty()) {
            int index = sta.top();
            sta.pop();
            int current = heights[index] * (i - (sta.empty() ? -1 : sta.top()) - 1);
            result = max(result, current);
        }
        return result;
    }
};